elasticai/creator/vhdl/skeleton_id.py

```diff
@@ -0,0 +1,146 @@
+"""
+The skeleton id should be computed and set after all other code is generated.
+"""
+
+import logging
+from collections.abc import Iterable
+from hashlib import blake2b, file_digest
+from importlib.metadata import version as _pypackage_version
+from itertools import filterfalse
+from pathlib import Path
+
+import tomlkit as toml
+from tomlkit.toml_file import TOMLFile
+
+
+def _is_id(line: str):
+    return line.strip().startswith("constant SKELETON_ID")
+
+
+def _build_skeleton_id_line(id: Iterable[str]) -> str:
+    id_str = ", ".join(f'x"{b.upper()}"' for b in id)
+
+    return f"  constant SKELETON_ID : skeleton_id_t := ({id_str});"
+
+
+def replace_id_in_code(code: Iterable[str], id: Iterable[str]) -> Iterable[str]:
+    """
+    Look for a line that starts with `constant SKELETON_ID` and replace it with
+    the given id.
+    """
+    for line in code:
+        if _is_id(line):
+            yield _build_skeleton_id_line(id)
+            yield "\n"
+        else:
+            yield line
+
+
+class SkeletonIdHash:
+    SIZE: int = 16
+
+    def __init__(self):
+        self._digests = []
+
+    def _hash(self):
+        return blake2b(digest_size=self.SIZE)
+
+    def update(self, lines: Iterable[str] | Path) -> None:
+        h = self._hash()
+
+        if isinstance(lines, Path):
+            with open(lines, "rb") as f:
+                self._digests.append(file_digest(f, self._hash).digest())
+        else:
+            for line in lines:
+                h.update(line.encode())
+            self._digests.append(h.digest())
+
+    def digest(self) -> bytes:
+        h = self._hash()
+        for d in sorted(self._digests):
+            h.update(d)
+        return h.digest()
+
+
+def compute_skeleton_id_hash(
+    build_dir: Path, file_with_default_id: str = "skeleton_pkg.vhd"
+) -> bytes:
+    logger = logging.getLogger(__name__)
+    hash = SkeletonIdHash()
+    logger.info("writing skeleton id to skeleton_pkg.vhd")
+    all_files = build_dir.glob("**/*.vhd")
+
+    def skip_default_id_for_idempotency(file: Path) -> bool:
+        return file.name == file_with_default_id
+
+    files_to_hash = filterfalse(skip_default_id_for_idempotency, all_files)
+    for vhd_file in files_to_hash:
+        logger.debug(f"hashing {vhd_file.as_uri}")
+        hash.update(vhd_file)
+    digest = hash.digest()
+    logger.debug(f"raw_digest is {digest.hex()}")
+
+    return digest
+
+
+def get_skeleton_pkg_file(build_dir: Path) -> Path:
```

Comment:
maybe make it a private function

Reply:
yep, especially when the filename is hard coded, not sure it should be. I was trying to get that feature in instead of being perfectionist about it ;)